plugins/woocommerce/client/admin/client/settings-payments/settings-payments-main.tsx

```diff
@@ -17,64 +21,129 @@ import { WooPaymentsGatewayData } from '~/settings-payments/types';
-		Plugin[]
-	>( [] );
 	const [ wooPaymentsGatewayData, setWooPaymentsGatewayData ] = useState<
 		WooPaymentsGatewayData | undefined
 	>( undefined );
 	const [ installingPlugin, setInstallingPlugin ] = useState< string | null >(
 		null
 	);
+	const [ isInstalled, setIsInstalled ] = useState< boolean >( false );
+	const [ isEnabled, setIsEnabled ] = useState< boolean >( false );
 	const { installAndActivatePlugins } = useDispatch( PLUGINS_STORE_NAME );
 
 	const installedPluginSlugs = useSelect( ( select ) => {
 		return select( PLUGINS_STORE_NAME ).getInstalledPlugins();
 	}, [] );
 
+	// Make UI to refresh when plugin is installed.
+	const { invalidateResolution } = useDispatch( 'core/data' );
+
 	useEffect( () => {
-		setWooPaymentsGatewayData(
-			parseScriptTag( 'experimental_wc_settings_payments_woopayments' )
-		);
-		setRegisteredPaymentGateways(
-			parseScriptTag( 'experimental_wc_settings_payments_gateways' )
-		);
-		setPreferredPluginSuggestions(
-			parseScriptTag(
-				'experimental_wc_settings_payments_preferred_extensions_suggestions'
-			)
-		);
-		setOtherPluginSuggestions(
-			parseScriptTag(
-				'experimental_wc_settings_payments_other_extensions_suggestions'
-			)
-		);
-	}, [] );
+		if ( isInstalled ) {
+			invalidateResolution(
+				PAYMENT_GATEWAYS_SUGGESTIONS_STORE_NAME,
+				'getPaymentGateways'
+			);
+		}
+	}, [ isInstalled, invalidateResolution ] );
+
+	const {
+		registeredPaymentGateways,
+		preferredPluginSuggestions,
+		otherPluginSuggestions,
+		isFetching,
+	} = useSelect( ( select ) => {
+		return {
+			registeredPaymentGateways: select(
+				PAYMENT_GATEWAYS_SUGGESTIONS_STORE_NAME
+			).getPaymentGateways(),
+			preferredPluginSuggestions: select(
+				PAYMENT_GATEWAYS_SUGGESTIONS_STORE_NAME
+			).getPreferredPluginSuggestions(),
+			otherPluginSuggestions: select(
+				PAYMENT_GATEWAYS_SUGGESTIONS_STORE_NAME
+			).getOtherPluginSuggestions(),
+			isFetching: select(
+				PAYMENT_GATEWAYS_SUGGESTIONS_STORE_NAME
+			).isFetchingPaymentGatewaySuggestions(),
+		};
+	} );
 
+	// test
 	const setupPlugin = useCallback(
-		( plugin: Plugin ) => {
+		( extension: SuggestedPaymentExtension ) => {
 			if ( installingPlugin ) {
 				return;
 			}
-			setInstallingPlugin( plugin.id );
-			installAndActivatePlugins( [ plugin.plugins[ 0 ] ] )
+			setInstallingPlugin( extension.id );
+			installAndActivatePlugins( [ extension.plugin.slug ] )
 				.then( ( response ) => {
 					createNoticesFromResponse( response );
-					// Reload the page to reflect the changes - will be replaced with a more elegant solution in the future.
-					window.location.reload();
+					setIsInstalled( true );
+					setInstallingPlugin( null );
 				} )
 				.catch( ( response: { errors: Record< string, string > } ) => {
 					createNoticesFromResponse( response );
 					setInstallingPlugin( null );
 				} );
 		},
-		[ installAndActivatePlugins, installingPlugin ]
+		[ installingPlugin, isInstalled ]
+	);
+
+	const togglePlugin = useCallback(
```

Comment:
@woocommerce/ghidorah I would appreciate a lot if you help me to extract it into store action properly so I can call it using useDispatch.
I tried, but window.woocommerce admin is not set, then we have redirect (window.location.href) in some cases - I am a bit confused how to do it properly.
What I've read is that action should update the store and then the UI will re-render if everything set up properly. Now I manually re-render it with `invalidateResolution` a bit above

Reply:
I think I got it - fixed in https://github.com/woocommerce/woocommerce/pull/52824/commits/7eb1dd7c189a8425cfd5345c6cfefe3e436c5783
let me know what you think!

Reply:
It's good!